### ISSUE TITLE
Pyright does not account for subtypes when inferring short-circuiting of type

### ISSUE THREAD

**Issue by gerhardadler** [microsoft/pyright#9938]
**Describe the bug**
Pyright is quite smart when it comes to inferring the type in short-circuit scenarios. It seems to check the signature of the `__bool__` method of the type to see how it will evaluate. This works fine in most schenarios, but when doing subclassing, some type errors are not caught.

Take this example:
```python
class C:
    def name(self) -> str:
        return 'I am C'

def f1(x: str | None) -> None: ...

def f2(c: C | None):
    f1(c and c.name())

class D(C):
    def __bool__(self) -> bool:
        return False

f2(D())
```
`f2` accepts `C`, and as a result also subtypes of `C`. It correctly infers that C will always be trueish, and therefore, it evaluates the expression `c and c.name()` to be `None | str`. 

However when creating a subclass `D` which overrides `__bool__` to return False, Pyright does not show an error when passing an instance of `D` to `f2`.

**VS Code extension or command-line**
I'm running Pyright in the Pyright playground.

Pyright version: 1.1.394
Python version: 3.12
Strict: both

**Comment by erictraut** [comment]
Pyright is working as intended here, so I don't consider this a bug.

You're correct that pyright doesn't catch an error in the case that you've highlighted. I wouldn't expect it to in this case.

The problem here is that the `bool()` check that Python performs at runtime to evaluate a conditional statement is a function that effectively violates the Liskov Substitution Principle (LSP). There are other edge cases where the standard library and Python runtime also unfortunately violate the LSP, but there's not much that static type checkers can do in these cases. The type system is built upon an assumption that the LSP is valid. If you specify the type `C` in a type annotation, the type system assumes that all subclasses of `C` will act just like `C` with regard to their public interface contract. 

There are cases where you can make your code more robust by working with static analysis tools and avoiding patterns and use cases that cannot be caught by them. For example, you could avoid using shortcut behaviors (which arguably make the code harder to read and understand). You could also define an explicit method or property as part of `C`'s interface rather than relying on `__bool__` (where the intent of the code is not clear, since it's just a truthy or falsy test).
